Warn when duplicate field labels are detected

ezraporter opened the pull request:
# Description
This PR implements a warning when `REDCapTidieR` detects a radio or dropdown field with duplicated labels. It also fixes the cryptic error noted in https://github.com/CHOP-CGTInformatics/REDCapTidieR/issues/89#issuecomment-1403937134

# Proposed Changes 
List changes below in bullet format:

- Create `check_parsed_labels()` check function and call it in `multi_choice_to_labels()` after the labels are parsed
    - I thought about doing this within `parse_labels()` but this function gets called multiple times
- Add a field with duplicate labels to the classic test REDCap and `suppressWarnings()` where necessary throughout the tests

**Screenshots**
Example warning:

![Screenshot 2023-01-26 at 2 49 21 PM](https://user-images.githubusercontent.com/60618324/214935805-7b2071bb-11f8-42a9-9f99-609c56bd5813.png)

### Issue Addressed
closes #89

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [x] New/revised functions that update downstream outputs have associated static testing files (`.RDS`) updated under `inst/testdata/create_test_data.R`
- [x] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR
- [x] Pre-release package version incremented using `usethis::use_dev_version()`

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

rsh52 commented on tests/testthat/test-read_redcap.R:
```diff
@@ -19,7 +19,8 @@ test_that("read_redcap works for a classic database with a nonrepeating instrume
         "field_missing_categories",
-        "empty_parse_warning"
+        "empty_parse_warning",
+        "duplicate_labels"
```
Is there a better way we could handle the growing number of warnings we have to account for in our tests? We can't just reference `classes = "REDCapTidieR_cond"`?

ezraporter replied:
Ha I was thinking this too. We can do that as long as we're fine silencing *other* `REDCapTidieR` warnings besides the ones we expect.

An alternative is just to move our warning-generating cases to a new REDCap so our main test REDCap will be warning-free. Not sure if that's better or worse

rsh52 replied:
Hm, it might pay to look into cleaning up our test databases like how Will has 50 for unique test cases. But that will also be a semi-large lift.

Not anything necessary for this PR but something for us to throw on the pile of considerations ✨

rsh52 reviewed (changes requested):
Overall this looks great, a few comments and suggestions related to tightening up naming conventions/documentation open to discussion.

ezraporter commented on R/checks.R:
```diff
@@ -254,6 +254,40 @@ check_req_labelled_metadata_fields <- function(supertbl, call = caller_env()) {
+#' @importFrom cli cli_warn qty
+#' @importFrom rlang caller_env
+#'
+#' @param parsed_labs a vector of parsed labels produced by `parse_labels()`
```
👍  Will implement this

ezraporter commented on R/checks.R:
```diff
@@ -254,6 +254,40 @@ check_req_labelled_metadata_fields <- function(supertbl, call = caller_env()) {
 
 
+#' @title
+#' Check that parsed labels are not duplicated
```
This makes sense to me. I'll rename to `check_parsed_labels_duplicates` and if we later add more general checks we can add a `check_parsed_labels` that either calls or incorporates `check_parsed_labels_duplicates`.

ezraporter commented on R/checks.R:
```diff
@@ -254,6 +254,40 @@ check_req_labelled_metadata_fields <- function(supertbl, call = caller_env()) {
+
+    msg <- c(
+      "!" = "Multiple values are mapped to the {qty(dups)} label{?s} {.code {dups}} in field {.code {field_name}}",
+      "i" = "Consider making the labels for {.code {field_name}} unique in your REDCap project"
```
Probably not but I'm unsure of the best way to fix that. I'd like to avoid a solution that requires the check to know whether or not we stripped html tags because this happens inside `parse_labels`. A simple alternative is to just always include a note saying "`REDCapTidieR` removes html tags from labels which may have caused the duplication."

What do you think?

rsh52 replied:
It depends how worthwhile we find alerting users to HTML. I'm just thinking in Roni's case the current error would have probably been confusing anyway since he/someone might have thought his labels were unique, but it was actually on our end that we made them duplicates.

It would be more work, and may not be worthwhile, but possibly we could expand `strip_html_field_embedding` to also `detect_html_field_embedding` and if triggered (or is `TRUE`, or whatever), append the `cli` message to include the note you suggested.

skadauke replied:
Discussed checking for empty values. If duplicate values are due to only empty values then only show warning related to empty labels and not also the duplicate label warning.

Add the following if this is because we removed HTML?
`"i" = "This may happen if the label only contains HTML"`

ezraporter commented:
Made the following updates:
- Added a `return_stripped_text_flag` parameter to `parse_labels()` to have it optionally return a flag indicating whether we changed the labels by stripping text
    - Minor refactoring of `parse_labels()` to achieve this
- Pass that flag to `check_parsed_labels()` to determine whether we should add the note about HTML
- Added a check to `check_parsed_labels()` to warn if any labels are blank ("")

rsh52 reviewed (approved):
This is definitely starting to get complicated, but I think it looks good and I can generally follow the qualifier variables/arguments informing what checks we expect to fire off.

The tests also look good, we just want to be mindful that some of the test cases we're addressing that prompted these functions are spread across smaller tests.
